crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -264,3 +163,479 @@ impl From<ClientNameRequestUnicodeFlag> for CharacterSet {
+    pub fn new_response(capabilities: Vec<CapabilityMessage>) -> Self {
+        Self {
+            num_capabilities: capabilities.len() as u16,
+            padding: 0,
+            capabilities,
+            kind: CoreCapabilityKind::ClientCoreCapabilityResponse,
+        }
+    }
+
+    pub fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        ensure_size!(ctx: "CoreCapability", in: dst, size: self.size());
+        dst.write_u16(self.num_capabilities);
+        dst.write_u16(self.padding);
+        for cap in self.capabilities.iter() {
+            cap.encode(dst)?;
+        }
+        Ok(())
+    }
+
+    pub fn decode(header: SharedHeader, payload: &mut ReadCursor<'_>) -> PduResult<Self> {
+        let kind = match header.packet_id {
+            PacketId::CoreServerCapability => CoreCapabilityKind::ServerCoreCapabilityRequest,
+            PacketId::CoreClientCapability => CoreCapabilityKind::ClientCoreCapabilityResponse,
+            _ => {
+                return Err(invalid_message_err!(
+                    "CoreCapability::decode",
+                    "PacketId",
+                    "invalid value"
+                ))
+            }
+        };
+
+        ensure_size!(ctx: kind.name(), in: payload, size: Self::FIXED_PART_SIZE);
+
+        let num_capabilities = payload.read_u16();
+        let padding = payload.read_u16();
+        let mut capabilities = vec![];
+        for _ in 0..num_capabilities {
+            capabilities.push(CapabilityMessage::decode(payload)?);
+        }
+
+        Ok(Self {
+            num_capabilities,
+            padding,
+            capabilities,
+            kind,
+        })
+    }
+
+    pub fn size(&self) -> usize {
+        Self::FIXED_PART_SIZE + self.capabilities.iter().map(|c| c.size()).sum::<usize>()
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum CoreCapabilityKind {
+    /// [2.2.2.7 Server Core Capability Request (DR_CORE_CAPABILITY_REQ)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/702789c3-b924-4bc2-9280-3221bc7d6797)
+    ServerCoreCapabilityRequest,
+    /// [2.2.2.8 Client Core Capability Response (DR_CORE_CAPABILITY_RSP)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/f513bf87-cca0-488a-ac5c-18cf18f4a7e1)
+    ClientCoreCapabilityResponse,
+}
+
+impl CoreCapabilityKind {
+    fn name(&self) -> &'static str {
+        match self {
+            CoreCapabilityKind::ServerCoreCapabilityRequest => "ServerCoreCapabilityRequest",
+            CoreCapabilityKind::ClientCoreCapabilityResponse => "ClientCoreCapabilityResponse",
+        }
+    }
+}
+
+/// [2.2.1.2.1 Capability Message (CAPABILITY_SET)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/f1b9dd1d-2c37-4aac-9836-4b0df02369ba)
+#[derive(Debug, Clone, Copy)]
+pub struct CapabilityMessage {
+    header: CapabilityHeader,
+    capability_data: CapabilityData,
+}
+
+impl CapabilityMessage {
+    const NAME: &str = "CapabilityMessage";
```

Comment:
Nit: let’s use the upper cased structure name from Microsoft documentation for consistency (this is what we’ll get in errors)
```suggestion
    const NAME: &str = "CAPABILITY_SET";
```
In some cases there is no such name, and in such cases the name of the Rust type is used instead (`ironrdp_pdu::geometry::InclusiveRectangle` for instance)

Reply:
001d937d366b978775b62aa9455105ebd60bfc6c